docs/blueprints/planned-downtime.md

```diff
@@ -0,0 +1,62 @@
+
+### Transferring Pieces
+
+Conveniently, we already have protobufs and example code for transferring pieces from one node to another as a result of (TODO link) [Graceful Exit](), so implementing this type of transfer should be fairly straightforward.
+
+The main thing that needs to be added on the storagenode side is an endpoint exclusively for satellite use that allows the satellite to initiate a piece transfer request unrelated to graceful exit. It is possible that we could use the same code path that graceful exit will use (we are currently refactoring graceful exit), but this might not be the best idea. In any case, it is worth considering, especially if we want to stream many piece transfer requests back-to-back.
+
+On the satellite side, we will need to add a chore that does the following:
+1. Get a list of nodes with upcoming planned downtime within a configurable interval (TODO add list of config values to this doc)
+2. Create a metainfo loop observer which generates a list of unhealthy and/or close-to-unhealthy segments which these planned downtime nodes are also a part of (can this somehow be added to the checker?). The output of the observer should be a map of node IDs to piece IDs which need to be transferred (as well as any other important info in metainfo needed for a piece transfer). (TODO how do we limit memory usage?)
+3. Create a limiter for opening concurrent connections to nodes, and using this limiter, for each node, attempt to transfer all of the relevant pieces.
+
+An attempt to transfer a stream of pieces for a given node will look like:
+1. Open connection to stream piece transfer requests to nodes (like graceful exit, but initiated by satellite)
+2. Select target node from overlay cache. Exclude nodes with upcoming planned downtime (TODO come up with easy to say name that describes "nodes with upcoming planned downtime")
+3. Construct and sign piece transfer request with origin node, target node, piece ID, any other relevant info
+4. Send piece transfer request to node and wait for response
+5. On failure, retry (TODO config value for limiting number of retries)
+6. On success, go to next piece and repeat from step 2
+7. On list completion, close connection
+
+The chore will run on a configured interval - if there are nodes with upcoming planned downtime, it will go ahead and try to transfer pieces off of them. Otherwise, it will wait before checking again.
+
+Open questions:
+
+* How do we limit planned downtime? I see two main options (but there are probably more): 1. Have a specific number of hours per year that a node can use for planned downtime. Once they go over, they simply stop being able to schedule planned downtime. 2. No strict limit on planned downtime. Rather, financially disincentivize nodes from taking planned downtime. No disincentive for first x hours, but after that, planned downtime can reduce payout
+* What should the interface for planned downtime look like?
```

Comment:
I have pictured something like:

* on the node console, I click "schedule planned downtime"
* I choose a time for downtime to start, between 24 hours and 1 week from now
* I say how long I plan for my node to be down
* I hit a submit button or something
* The console shows something like "Satellite is determining pieces to transfer off of this node. Please check back in (at most) X hours."
* It might also show something like "Transfer progress: 25/521 (4%)   |   Successful: 24   |   Failed: 1"
* After all pieces are done transferring (or like 80% of them? I dunno) it says "Downtime planning successful. Planned downtime in: 2 days 4 hours 32 minutes"

Reply:
Interesting. I was thinking something like:
* on the node console, there is a message "no upcoming planned downtime" or "upcoming planned downtime at xxxx-xx-xx, xx:xx UTC" with a button "add/update upcoming planned downtime" - the interface only allows you to schedule if more than a week in advance or something (maybe like 3 days? But I feel like 24 hours might not be enough time)
* Then, the satellite is free to send piece transfer requests to the node, which it responds to in the same way as any other request - it does the work and saves the orders so that it can get paid. I didn't imagine having a concept of "transfer progress". I'm not opposed to it, but I'm curious if you think it is necessary to design it in that way